### ISSUE TITLE
LogServerSignatureResult.Invalid.SignatureFailed

### ISSUE THREAD

**Issue by kSriramji** [Babylonpartners/certificate-transparency-android#28]
Started getting **Invalid Signature**  for the same domain which was working fine until yesterday.

**Comment by mattmook** [comment]
Both should work, although we should all migrate to the v2 schema. I don't know if/when the v1 schema will be removed.

Indeed with the latest release of this library, v0.2.0, we have removed support for the v1 schema completely.


What version of the library are you using and what is the domain you are hitting so I can try to replicate this.

**Comment by pholst** [comment]
I experienced same issue earlier today:`Unable to load log servers with SignatureVerificationFailed(signatureResult=Invalid signature)` while I was running `certificatetransparency-android:0.1.4`

However, it did suddenly work again around 2019-08-22 00:14:15 CET.

Have any log server been unavailable today?

**Comment by kSriramji** [comment]
@mattmook As @pholst mentioned we are also using the same version certificatetransparency-android:0.1.4

You may be aware I have raised issue in google group as well after the discussion it started working we have no idea about the **root cause yet**.

Ref : https://groups.google.com/d/msg/certificate-transparency/zZwGExvQeiE/3GAOK2WvDgAJ

**Comment by mattmook** [comment]
I saw the discussion in the Google Group, there are a couple of things that might have happened.

firstly at a point in time the json and sig files may have been out of sync. we don't know how long this happened for if it even happened. with that regard, there's not much we can do about that.

secondly though, the library could have inadvertantly cached one of those files out of sync with the other. any issue could have persisted for longer than the files were actually out of sync. as such I'm going to re-evaluate how caching is working for these files and ensure there are suitable tests to reduce issues like this and ensure if they do occur they don't persist for long.

**Comment by kSriramji** [comment]
@mattmook Thanks for the evaluation. 

I'm afraid this can be applicable for updated library **certificatetransparency-android:0.2.0**(version 2 schema) also

Note: We have tested the incident more than 4 different devices & everything gives me the same result.

**Comment by blinkmacalahan** [comment]
I recently added this library and things were working fine. Starting today, cert transparency is failing:

Failure: Unable to load log servers with SignatureVerificationFailed(signatureResult=Invalid signature)

**Comment by maxkohne** [comment]
we are sporadically getting the same thing too using `0.2.0`

`
 Failure: Unable to load log servers with SignatureVerificationFailed(signatureResult=Invalid signature)
`

**Comment by blinkmacalahan** [comment]
> I recently added this library and things were working fine. Starting today, cert transparency is failing:
> 
> `Failure: Unable to load log servers with SignatureVerificationFailed(signatureResult=Invalid signature)`
> 
> I'm using version `0.2.0` of the library.
> 
> Looking further, it seems to be an issue with the `Signature` class failing verification of the log_list.json . I found this [thread](https://groups.google.com/d/msg/certificate-transparency/zZwGExvQeiE/Ye2IbxO1DgAJ) and I'm having issues with the exact thing.

This issue is no longer happening. It's very scary that one moment it works and then another it doesn't.

**Comment by maxkohne** [comment]
> > I recently added this library and things were working fine. Starting today, cert transparency is failing:
> > `Failure: Unable to load log servers with SignatureVerificationFailed(signatureResult=Invalid signature)`
> > I'm using version `0.2.0` of the library.
> > Looking further, it seems to be an issue with the `Signature` class failing verification of the log_list.json . I found this [thread](https://groups.google.com/d/msg/certificate-transparency/zZwGExvQeiE/Ye2IbxO1DgAJ) and I'm having issues with the exact thing.
> 
> This issue is no longer happening. It's very scary that one moment it works and then another it doesn't.

Exact same thing happened to us. It came and then went. This is the second time we randomly saw a sudden spike in errors. We have debated whether to take out this library of our app but we don't want to sacrifice security. We are going to do that as a very last resort. After this second spike in incidents, we have decided to have a way to remotely enable or disable this feature.

Here is an example of our Kotlin code for our app:


```
Interceptor certificateTransparencyInterceptor = new Interceptor() {
    private Interceptor _certificateTransparencyInterceptor =
        new CTInterceptorBuilder()
            .includeHost("*.example.com")
            .setLogger(new BasicAndroidCTLogger(BuildConfig.DEBUG))
            .build();

    @NonNull
    @Override
    public Response intercept(@NonNull Chain chain) throws IOException {
        // FeatureFlag is a class we made that wraps a remote feature flag toggle service
        if (FeatureFlag.CERTIFICATE_TRANSPARENCY.isEnabled()) {
            return _certificateTransparencyInterceptor.intercept(chain);
        } else {
            return chain.proceed(chain.request());
        }
    }
};

// Some example client
OkHttpClient exampleOkHttpClient = getClient().newBuilder()
    .addNetworkInterceptor(certificateTransparencyInterceptor)
    .build();
```

**Comment by blinkmacalahan** [comment]
As I previously mentioned, I recently added this library and when doing so, I added logs with `CTLogger` to capture `NoCertificates, is LogServersFailed, NoScts, is TooFewSctsTrusted, is UnknownIoException` in our codebase. These logs are sent to our application monitor. After encountering this issue, I reviewed my logs and found the same issue has happened a few other times:

- Jan 22, 2020 7:50:49 AM UTC
- Jan 30, 2020 10:17:36 AM UTC

It has occurred on various devices and OS versions.

**Comment by maxkohne** [comment]
We saw the same spike on Jan 30 and Feb 5. Interesting!

**Comment by vivascu** [comment]
Having faced the same issue today our investigation concluded that the problem was on the wrong certificate transparency log was served by [https://www.gstatic.com/ct/log_list/v2/log_list.json ](https://www.gstatic.com/ct/log_list/v2/log_list.json ).

We described the issue here https://github.com/google/certificate-transparency/issues/1476 if you want to go in detail.
**TLDR** The library is requesting le log list with a header `Accept-Encoding: Gzip` which gives back a cached version of the log list. 

The cache expired in 24 hours and now things are in their place. I imagine this will happen again. We could either explicitly remove the header added in the library or rely on the log list to be next time updated  simultaneously for both cases, or just wait for the cache to expire. 

https://github.com/babylonhealth/certificate-transparency-android/pull/39 is a welcomed addition I don't think it fixes the problem.

**Comment by jwagner** [comment]
@vivascu Removing the header is not enough. The non gzipped "version" is also cached and can also diverge from the .sig file. Right now fetching the signature and json from gstatic is fundamentally broken because their cache lifetimes are not the same.

**Comment by blinkmacalahan** [comment]
It looks like Google has attempted to fix the caching problem by creating a zip file containing the files: https://github.com/google/certificate-transparency/issues/1476. @mattmook I was curious if the library needs to be updated to use this zip or not.

**Comment by mattmook** [comment]
@blinkmacalahan yes it will require an update, I’ve got a PR ready but just doing some more testing on it first before I hit merge. https://github.com/babylonhealth/certificate-transparency-android/pull/49
